stdlib/public/Concurrency/TaskGroup.cpp

```diff
@@ -964,16 +965,37 @@ static void swift_taskGroup_initializeImpl(TaskGroup *group, const Metadata *T)
 SWIFT_CC(swift)
 static void swift_taskGroup_initializeWithFlagsImpl(size_t rawGroupFlags,
                                                     TaskGroup *group, const Metadata *T) {
-
-#if !SWIFT_CONCURRENCY_EMBEDDED
-  ResultTypeInfo resultType;
-  resultType.metadata = T;
-
   TaskGroupFlags groupFlags(rawGroupFlags);
   SWIFT_TASK_GROUP_DEBUG_LOG_0(group, "create group, from task:%p; flags: isDiscardingResults=%d",
                                swift_task_getCurrent(),
                                groupFlags.isDiscardResults());
 
+  ResultTypeInfo resultType;
+#if !SWIFT_CONCURRENCY_EMBEDDED
+  resultType.metadata = T;
+  assert(!groupFlags.isMetadataAsOptionRecord());
+#else
+  assert(groupFlags.isMetadataAsOptionRecord());
+  TaskOptionRecord *options = (TaskOptionRecord *)T;
```

Comment:
Can't say I like this if I'm honest... Feels like a confusing trickery we'll end up paying for someday.

> The runtime entry point (swift_taskGroup_initializeWithFlags) doesn't offer a convenient way to pass extra data (unlike swift_task_create). I'm avoiding adding a new entry for simplicity and instead using a bit in groupFlags that means that instead of Metadata * we receive the same format of arbitrary options that swift_task_create uses (TaskOptionRecord linked list).

When we needed options in task creation we _added_ new API to handle it. Wouldn't it be more proper to do the same here?

WDYT @rjmccall ?

Reply:
I agree.  It's not like this kind of trick makes the code work with an existing runtime anyway — you have to test that the runtime supports the fancy new thing you're doing, so you might as well call a new entrypoint.

Reply:
Added the new entrypoint.